pageserver/src/relish_storage.rs

```diff
@@ -0,0 +1,103 @@
+//! Abstractions for the page server to store its relish layer data in the external storage.
+//!
+//! Main purpose of this module subtree is to provide a set of abstractions to manage the storage state
+//! in a way, optimal for page server.
+//!
+//! The abstractions hide multiple custom external storage API implementations,
+//! such as AWS S3, local filesystem, etc., located in the submodules.
+
+mod local_fs;
+mod rust_s3;
+/// A queue and the background machinery behind it to upload
+/// local page server layer files to external storage.
+pub mod storage_uploader;
+
+use std::{
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+/// Storage (potentially remote) API to manage its state.
+#[async_trait::async_trait]
+pub trait RelishStorage: Send + Sync {
```

Comment:
Thoughts on having `type RelishStoagePath;`, with each `impl RelishStorage` defining an appropriate one for their purposes? This would make the uploader generic over RelishStorage.